src/combobox/ComboBox.ts

```diff
@@ -0,0 +1,315 @@
+			this.invalidate();
+		},
+
+		Enter(this: ComboBox) {
+			const { results } = this.properties;
+
+			if (!this._open || !results || results.length === 0) {
+				return;
+			}
+			// If no result is highlighted, close menu
+			else if (this._selectedIndex === undefined) {
+				this._open = false;
+				this.invalidate();
+			}
+			// Select the highlighted result
+			else {
+				this.selectResult(this.getResultValue(results[this._selectedIndex]));
+			}
+		},
+
+		Escape(this: ComboBox) {
+			// Close menu
+			this._open = false;
+			this._selectedIndex = undefined;
+			this.invalidate();
+		}
+	};
+
+	afterCreate(element: HTMLElement) {
+		// Cache the input element and restore focus
+		this._inputElement = element.querySelector('input');
+		this._inputElement![this._focused ? 'focus' : 'blur']();
+	}
+
+	afterUpdate(element: HTMLElement) {
+		// Cache the input element and restore focus
+		this._inputElement![this._focused ? 'focus' : 'blur']();
+		const selectedResult = element.querySelector('[data-selected="true"]');
+		// Make sure highlighted result is visible after arrow up / arrow down
+		selectedResult && this.scrollIntoView(<HTMLElement> selectedResult);
+	}
+
+	scrollIntoView(element: HTMLElement) {
+		const menu = <HTMLElement> element.parentElement;
+		// Scroll menu up so top of highlighted result aligns with top of menu container
+		if (element.offsetTop - menu.scrollTop < 0) {
+			menu.scrollTop = element.offsetTop;
+		}
+		// Scroll menu down so bottom of highlighted result aligns with bottom of menu container
+		else if ((element.offsetTop - menu.scrollTop + element.offsetHeight) > menu.clientHeight) {
+			menu.scrollTop = element.offsetTop - menu.clientHeight + element.offsetHeight;
+		}
+	}
+
+	onInput(event: Event) {
+		const {
+			onRequestResults,
+			onChange
+		} = this.properties;
+
+		// Open the menu and request new results as typing occurs
+		this._selectedIndex = undefined;
+		this._open = true;
+		onChange && onChange((<HTMLInputElement> event.target).value);
+		onRequestResults && onRequestResults((<HTMLInputElement> event.target).value);
+	}
+
+	onInputFocus(event: FocusEvent) {
+		const {
+			openOnFocus,
+			onFocus,
+			onRequestResults
+		} = this.properties;
+
+		const value = (<HTMLInputElement> event.target).value;
+
+		this._focused = true;
+		this._ignoreBlur = false;
+		onFocus && onFocus(value);
+
+		if (openOnFocus && !this._ignoreFocus) {
+			this._open = true;
+			onRequestResults && onRequestResults(value);
+		}
+
+		this._ignoreFocus = false;
+	}
+
+	onInputBlur(event: FocusEvent) {
+		const { onBlur } = this.properties;
+
+		if (this._ignoreBlur) {
+			return;
+		}
+
+		this._open = false;
+		this._focused = false;
+		this._selectedIndex = undefined;
+		onBlur && onBlur((<HTMLInputElement> event.target).value);
+		this.invalidate();
+	}
+
+	onInputKeyDown(event: KeyboardEvent) {
+		this._handlers[event.key] && this._handlers[event.key].call(this, event);
```

Comment:
This still isn't supported in Safari (desktop and iOS), sadly :( http://caniuse.com/#feat=keyboardevent-key